lib/spack/spack/detection/path.py

```diff
@@ -241,15 +241,19 @@ def by_executable(
         path_hints: list of paths to be searched. If None the list will be
             constructed based on the PATH environment variable.
     """
+    no_path_hints = path_hints is None
     path_hints = spack.util.environment.get_path("PATH") if path_hints is None else path_hints
```

Comment:
I think this line could be rewritten more clearly as:

```
if not path_hints:
  path_hints = spack.util.environment.get_path("PATH")
```

and actually it looks like regardless of whether we pull from `PATH`, if `path_hints` was originally empty, then we include  `common_windows_package_paths`, so we can also fold that in like:

```
if not path_hints:
  path_hints = spack.util.environment.get_path("PATH")
  path_hints.extend(common_windows_package_paths())
```

Reply:
Outside of the concern re the conditional I raised above, this looks good.

Reply:
Ok, I thought that might have been deliberate. I think we can still make this check:
```
if no_path_hints:
  path_hints = spack.util.environment.get_path("PATH")
  path_hints.extend(common_windows_package_paths())
```
and just make the `path_hints is None` check once.

Reply:
Ok, seems like mypy is still unable to follow control flow so these checks will now be:
```
    if path_hints is None:
        path_hints = spack.util.environment.get_path("PATH") if path_hints is None else path_hints
        path_hints.extend(common_windows_package_paths())
        no_path_hints = True
```
Which I think achieves what everyone is looking for here.

Reply:
@alalazo

> detection.by_executable(["gcc", "cmake"], path_hints=[])
> in which case we don't search anything (since the hint is explicitly empty).

I like to avoid that type of logic whenever possible: forcing users to distinguish between empty/None in Python is a source of confusion. If we want to be able to disable defaults in the presence of `[]`, then IMO we should have a `disable_defaults` parameter. However, my assessment is that we wouldn't want that, because a combined `path_hints=[], disable_defaults=True)` would lead to an empty set of search paths on Linux (i.e. we don't look in any paths, so we shouldn't be calling this function in the first place).

I can say that none of the call sites in core that I can find depend on this distinction:

```
lib/spack/spack/cmd/external.py:    detected_packages = spack.detection.by_executable(pkg_cls_to_check, path_hints=args.path)
lib/spack/spack/bootstrap/core.py:    externals = spack.detection.by_executable(search_list)
lib/spack/spack/build_systems/python.py:        python_externals_detection = spack.detection.by_executable(
```

Reply:
> I like to avoid that type of logic whenever possible: forcing users to distinguish between empty/None in Python is a source of confusion.

Hmm, I think it is a common idiom used in many places - also in the standard library. This is from the ProcessPoolExecutor constructor:
```python
        if max_workers is None:
            self._max_workers = os.cpu_count() or 1
            if sys.platform == 'win32':
                self._max_workers = min(_MAX_WINDOWS_WORKERS,
                                        self._max_workers)
        else:
            if max_workers <= 0:
                raise ValueError("max_workers must be greater than 0")

```
where they distinguish between `0` (not allowed) and `None`. The idiom is so common that `mypy` has an `Optional[X]` type hint that means the value might be type `X` or `None` , see [here](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html).

> I can say that none of the call sites in core that I can find depend on this distinction:

This in my opinion doesn't matter. We need a way to distinguish when to use a default value from PATH and when the client passed an empty list. Even from a common sense point of view:
```
specs = by_path(["cmake"])  # cmake found in PATH, not unexpected
specs = by_path(["cmake"], path_hints=[])  # cmake found in PATH ? Let's give strictly MORE things to search
specs = by_path(["cmake"], path_hints=["/tmp/foo"])  # cmake not found ? 
```

Reply:
Note that

`specs = by_path(["cmake"], path_hints=[])`

(with the semantics you desire) will never return anything on Linux, it is effectively a noop (so this type of call should never occur explicitly).

That being said, you are likely referring to cases where you compose `path_hints` programmatically. I'd say this creates confusion when you want to use default paths:

```
computed_path_hints = function() # let's say this is an empty list
# now I need to convert computed_path_hints to None explicitly if I want `by_executable` to find anything
```

I think that putting a `disable_defaults` parameter in the method signature is a clearer way of expressing this.

> I think it is a common idiom used in many places

That does not mean it's the right way to handle it here. I'm not planning a crusade on abuse of null values in our code base, but since I'm here making other requests in the vicinity of this code, I want to change this too.

Reply:
> Note that ... never occur explicitly

No, but they could occur when the value is read from another function:
```python
specs = by_path(["cmake"], path_hints=list_of_paths_to_search())
```

> I think that putting a disable_defaults parameter in the method signature is a clearer way of expressing this.

That's exactly what `None` means idiomatically in Python - I think it is a widely accepted idiom, while it's discouraged to use _falsey_ values, as they can be anything that can be casted to False. What I mean is that I did that change on purpose :slightly_smiling_face:

Reply:
I feel like this response ignores everything I said after

>> (with the semantics you desire) will never return anything on Linux, it is effectively a noop (so this type of call should never occur explicitly).

for example

> No, but they could occur when the value is read from another function:

I specifically anticipated this argument and addressed it.

>> I think that putting a disable_defaults parameter in the method signature is a clearer way of expressing this.

> That's exactly what None means idiomatically in Python

Yes, and I'm not a fan (the example I gave in https://github.com/spack/spack/pull/39850#discussion_r1321824504 demonstrates why). I'm worried about abuse of null; you are worried (possibly) about abuse of false-yness; my proposal to use `disable_defaults` addresses both (or at least, my proposal does not *require* abuse of false-yness).

Reply:
> Yes, and I'm not a fan (the example I gave in https://github.com/spack/spack/pull/39850#discussion_r1321824504 demonstrates why).

I don't think this example demonstrates it, as:
```
# Now I need to convert computed_path_hints to None explicitly if I want `by_executable` to find anything
```
implies that a user expects something to happen if an empty list is passed explicitly, which I think should not be the case. 

> my proposal to use disable_defaults addresses both (or at least, my proposal does not require abuse of false-yness).

In my opinion this adds an unnecessary argument, but I don't want to hold this PR on this. I think we just need to add a line in the docstring, to say:
```
If path_hints is None the paths are computed using defaults depending on the current platform.
```
or something similar.

Reply:
As an additional point, this is the code before #39843:

https://github.com/spack/spack/blob/7631b5ea14fb48f8859015b48fc379e06c9d4141/lib/spack/spack/detection/path.py#L233-L245

This behavior (which I think is correct) was introduced back in #33021 Before we had the _falsey_ kind of check.  

Proposal: can't we keep the optional behavior as it is in this PR and debate the internal API in another PR ?

Reply:
> Proposal: can't we keep the "optional" behavior as it is in this PR and debate the internal API in another PR ?

Sorry to chime in, but FWIW I agree this is somewhat outside the scope of this PR and seems to be a higher level debate than this PR was meant to be. If we could get this merged (both so we can re-enable those skipped tests and to support `Intel-MPI` detection on Windows (compiler wrapper detection via .bat)) and then I can make another PR to continue this conversation?